src/python/txtai/pipeline/text/translation.py

```diff
@@ -74,22 +76,28 @@ def __call__(self, texts, target="en", source=None):
         languages = self.detect(values) if not source else [source] * len(values)
         unique = set(languages)
 
-        # Build list of (index, language, text)
-        values = [(x, lang, values[x]) for x, lang in enumerate(languages)]
+        # Build a dict from language to list of (index, text)
+        langdict = defaultdict(list)
+        for x, lang in enumerate(languages):
+            langdict[lang].append((x, values[x]))
 
         results = {}
         for language in unique:
-            # Get all text values for language
-            inputs = [(x, text) for x, lang, text in values if lang == language]
+            # Get all indices and text values for a language
+            inputs = langdict[language]
```

Comment:
If `defaultdict` was selected due to the key possibly not being there, this could instead be `langdict.get(language)`

Reply:
I think the language is guaranteed to always be there because we construct ```langdict``` using all the languages

```
        for x, lang in enumerate(languages):
            langdict[lang].append((x, values[x]))
```

and afterward we are just looping on the unique set of languages again derived from ```languages```

```
        unique = set(languages)

```